### PR TITLE
Add governor api call provision to send cluster telemetry data

### PR DESCRIPTION
**Pull request by snootan**
Added http api call support to send cluster informations along with all workloads to VAC.
Added logic to get replica sets informations along with others.

### CODE REVIEW

File: src/pkg/data/consumers/governor/exporter.go
Comment (vikrantest) [inline src/pkg/data/consumers/governor/exporter.go:42]: nit: Nil checks in nested structures are preferred to avoid runtime nil pointer exceptions.

**Comment by 4everming** [comment]
It would be much better to include some codes for the test as well. The test can be triggered automatically by GitHub actions.  Here are some examples of unit and E2E tests: https://github.com/vmware-tanzu/cloud-native-security-inspector/blob/main/src/test/e2e/inspectionpolicy_test.go
https://github.com/vmware-tanzu/cloud-native-security-inspector/blob/main/src/pkg/policy/network/enforcer_test.go

File: src/api/v1alpha1/inspectionpolicy_types.go
Comment (bharathbhalu) [inline src/api/v1alpha1/inspectionpolicy_types.go:131]: In the case Enabled=true, are we validating Other fields are not empty?

If we can't do conditional validation base on Enabled using KubeBuilder. Then at least we need handle these checks at code level. This can cause crashloopback while running

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [inline src/pkg/inspection/controller.go:338]: Check whether  ApiUrl, ClusterId & ApiToken are not empty. if any one of them is empty, Throw the error.

File: src/pkg/data/consumers/governor/exporter.go
Comment (bharathbhalu) [inline src/pkg/data/consumers/governor/exporter.go:33]: In the case of Error, Throw Error with Right Msg(Governor Export Error). As in main loop we are throwing Controller Run Error
Also, add an Error log.

File: src/pkg/data/consumers/governor/exporter.go
Comment (bharathbhalu) [inline src/pkg/data/consumers/governor/exporter.go:40]: As this is just payload generation(Report Transformation) and nothing to do with API Request.
change it to getGovernorPayload.

File: src/config/samples/policy.yaml
Comment (bharathbhalu) [inline src/config/samples/policy.yaml:63]: Is the exporter handles both https and Http?

File: src/api/v1alpha1/workload.go
Comment (bharathbhalu) [inline src/api/v1alpha1/workload.go:44]: We don't need to capture ReplicaSet, we just need to capture Replica's count.

Comment (snootan) [reply src/api/v1alpha1/workload.go:44]: Intention was to capture whole info regarding replicaset. So, anyone could use in future if required. But if we required only count then I can change it.

Comment (bharathbhalu) [reply src/api/v1alpha1/workload.go:44]: During our debug session, we have seen that generated(KubeBuilder) `goharbor.io_assessmentreports.yaml` has increased by 2000 lines just by ReplicatSet spec.

Also in practical scenarios, I don't see any added value in Capturing a complete ReplicaSet.  Atleast I don't see value w.r.t Governor.

Also please update the  goharbor.io_assessmentreports.yaml with generated values using KubeBuilder.

File: src/config/samples/policy.yaml
Comment (snootan) [reply src/config/samples/policy.yaml:63]: Needs to be discussed with Engineering productivity team.

**Comment by JingChen23** [comment]
The failure of the cypress test is a known issue and has been fixed by https://github.com/vmware-tanzu/cloud-native-security-inspector/pull/176, please merge develop-0.3 and trigger the re-test.

Also,  let me change the base branch of this pr to develop-0.3, currently we should not file pr to the main branch.

**Comment by snootan** [comment]
Team, we have added go-client generated files temporarily. It will get removed as soon as client package will get published somewhere.

File: src/pkg/data/consumers/governor/exporter_test.go
Comment (bharathbhalu) [inline src/pkg/data/consumers/governor/exporter_test.go:37]: Write test cases for -ve scenarios for SendReportToGovernor like TimedOut and Invalid url.

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [inline src/pkg/inspection/controller.go:375]: Missing UnitTest for exportReportToGovernor

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [inline src/pkg/inspection/controller.go:337]: Missing UT for this block

File: src/pkg/data/consumers/governor/exporter.go
Comment (bharathbhalu) [inline src/pkg/data/consumers/governor/exporter.go:49]: This check is redundant

If nsa.Namespace.Name  is empty, then kubernetesWorkloads.Namespace = nsa.Namespace.Name results to empty.

And defualt value of kubernetesWorkloads.Namespace is also empty.


You can just replace it with `kubernetesWorkloads.Namespace = nsa.Namespace.Name`

File: src/pkg/data/consumers/governor/exporter.go
Comment (bharathbhalu) [inline src/pkg/data/consumers/governor/exporter.go:42]: Even can be a method on GovernorExporter. 

`func (g GovernorExporter) getGovernorAPIPayload(doc api.AssessmentReport) openapi.KubernetesTelemetryRequest 

can be called just by g. getGovernorAPIPayload ()`

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [inline src/pkg/inspection/controller.go:340]: We can be more specific with exact error.
`
if governorConfig.ClusterID == "" {
log.Error("GovernorExporter: Missing  ClusterID")
} 

if governorConfig.URL == "" {
} 

if governorConfig.APIToken == "" {
}
`

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [inline src/pkg/inspection/controller.go:341]: Log messages have to be more specific indicating GovernorExporter.

Rephrase all the error messages with the `Governor` prefix
log.Errorf("GovernorExporter : Error message")

File: src/pkg/inspection/controller.go
Comment (bharathbhalu) [reply src/pkg/inspection/controller.go:376]: There was no existing test case for the controller. We can skip it.

**Comment by snootan** [comment]
@bharathbhalu, UT parts will get covered in further iterations. And other changes have been done as part of csp implementation PR.

File: src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml
Comment (bharathbhalu) [inline src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml:148]: I think we have update governor properties at /src/tools/installation/yaml/crd.yaml
Can you check what is the impact when we don't add governor properties at crd.yaml

File: src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml
Comment (bharathbhalu) [inline src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml:150]: Also, update the governor properties in e2e tests.
/src/test/e2e/inspectionpolicyinspectionpolicy.yaml

Comment (bharathbhalu) [reply src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml:150]: Check how to run e2e test.

**Comment by bharathbhalu** [comment]
Have we tested all the workload kinds?
[ Deployment, StatefulSet, ReplicaSet, DaemonSet, Job, CronJob, Pod ]

please verify all the kinds possible.

File: src/pkg/data/consumers/governor/mock_cluster_client_api.go
Comment (pgalache) [reply src/pkg/data/consumers/governor/mock_cluster_client_api.go:1]: I wouldn't overcomplicate at this point. The updateTelemetry endpoint will hardly ever change so I think it's fine to generate mocks in case it changes and just add them in the PR.

We currently we are including go client model classes directly so I think it's fine to include mocks too. If at some point we import go client from somewhere, then generating mock clases would make more sense.

**Comment by vikrantest** [comment]
MR looks good to me know and also looks like that its ready for merge, i don't see any blocker in the workflows as part of this prototype integration.

File: src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml
Comment (snootan) [reply src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml:143]: This is generated from kube-builder.

File: src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml
Comment (snootan) [reply src/config/crd/bases/goharbor.goharbor.io_assessmentreports.yaml:148]: Its working fine with governor properties.

File: src/pkg/data/consumers/governor/mock_cluster_client_api.go
Comment (bharathbhalu) [reply src/pkg/data/consumers/governor/mock_cluster_client_api.go:1]: ok, then we can go with out auto mock generation.

**Comment by bharathbhalu** [comment]
@snootan PR is still in the Draft stage. We cant approve it when it's in the draft stage.

File: src/pkg/inspection/controller.go
Comment (4everming) [inline src/pkg/inspection/controller.go:147]: Why do we send an empty report to the governor when no namespace label selector matches?

Comment (snootan) [reply src/pkg/inspection/controller.go:147]: **Scenario is**: Suppose user has given label to one namespace and after sometime, if user has removed that label. Next time, agent will not found any namespace to scan. But, governor will be still showing the telemetry data to user. So, to remove discrepancy in data from both side, we did.

File: src/pkg/inspection/scanner.go
Comment (JingChen23) [inline src/pkg/inspection/scanner.go:164]: This function now does 2 things: get owner and the replica count, so can we change the name of the function to make it clearer? Or, is it better to seperate into 2 functions?

Comment (JingChen23) [reply src/pkg/inspection/scanner.go:164]: Anyway, this need to be resolved when we merge the code to main branch in the future.

WRT merging this change to feature-governer, now it is ok to write like that 😄

**Review by JingChen23 — APPROVED** [review]

**Pull request merged**
